File: solution/380. Insert Delete GetRandom O(1)/380.cpp
class RandomizedSet {
public:
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if (m_.count(val)) return false;
        
        nums_.push_back(val);
        m_[val] = nums_.size() - 1;
        return true;
    }
    
    bool remove(int val) {
        if (!m_.count(val)) return false;
        
        int new_pos = m_[val];
        int last_val = nums_.back();
        
        m_[last_val] = new_pos;
        nums_[new_pos] = last_val;
        nums_.pop_back();
        m_.erase(val);
        return true;
    }
    
    int getRandom() {
        return nums_[rand() % nums_.size()];
    }
    
private:
    vector<int> nums_;
    unordered_map<int, int> m_; // {val: index in nums_}
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: solution/380. Insert Delete GetRandom O(1)/380.py
class RandomizedSet:

    def __init__(self):
        self.A = []
        self.pos = {}

    def insert(self, val: int) -> bool:
        if val in self.pos:
            return False
        
        self.A.append(val)
        self.pos[val] = len(self.A) - 1
        return True

    def remove(self, val: int) -> bool:
        if val not in self.pos:
            return False
        
        i = self.pos[val]
        last = self.A[-1]
        self.A[i] = last
        self.pos[last] = i
        self.A.pop()
        del self.pos[val]
        return True

    def getRandom(self) -> int:
        return self.A[random.randint(0, len(self.A) - 1)]


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
